[llvm][LICM] Limit multi-use BOAssociation to FP and Vector

thetheodor opened the pull request:
Limit the re-association of BOps with multiple users to FP and Vector arithmetic.

thetheodor committed:
[LICM] Limit association of BinOps with multiple uses to FP and Vec types.

Reassociation and hoisting of binary ops with multiple uses can lead to
increased instruction count (and higher register pressure in some
targets).  This patch limits reassociation of binary ops with multiple
usersto FP and Vec types.

artagnon commented on llvm/lib/Transforms/Scalar/LICM.cpp:
```diff
@@ -2850,7 +2855,7 @@ static bool hoistBOAssociation(Instruction &I, Loop &L,
   auto *BO0 = dyn_cast<BinaryOperator>(BO->getOperand(LVInRHS));
   if (!BO0 || BO0->getOpcode() != Opcode || !BO0->isAssociative() ||
-      BO0->hasNUsesOrMore(3))
+      BO0->hasNUsesOrMore(BOAssociationUserLimit + 1))
```
Hm, isn't this actually a lower-bound (N Uses Or More)? Your description "Limit the number of users ..." doesn't seem to align. Also, what is the motivation behind this change?

thetheodor replied:
`hasNUsesOrMore(N+1)` itself is indeed a lower bound. However, in this case, if it's true then (next line) we don't perform the re-association. We are essentially checking if the number of uses is at most `N`. 

I could rewrite this to `!BO0->getNumUses() <= BOAssociationUserLimit` to make it clearer.

thetheodor replied:
Regarding the motivation: we (at NVIDIA) have found that in some cases too much re-association can lead to higher register pressure, especially when there are multiple uses inside a loop. We'd like a way to control this behavior.

fhahn replied:
Is LICM the correct place to fix this though? LICM is some form of canonicalization, w/o any consideration for register pressure. Is there any reason the backend cannot sink the instructions back if beneficial?

thetheodor replied:
In principle it makes sense to fix such RP issues in the backend. However, in our case, the the re-associated and hoisted instructions are further modified by other passes (LSR etc) and sinking them becomes non-trivial.

thetheodor commented on llvm/lib/Transforms/Scalar/LICM.cpp:
```diff
@@ -146,6 +146,11 @@ static cl::opt<unsigned> IntAssociationUpperLimit(
         "Set upper limit for the number of transformations performed "
         "during a single round of hoisting the reassociated expressions."));
 
+static cl::opt<unsigned> BOAssociationUserLimit(
```
I've included a test

nikic reviewed (commented):
So generally, what this transform is doing for the multi-use case is somewhat iffy. It goes against our usual policy to not increase instruction count.

This already came up in the discussion on the original PR in https://github.com/llvm/llvm-project/pull/81608 and https://github.com/llvm/llvm-project/pull/102114 later restricted it to at most two uses.

Ideally we'd find a way here to provide good behavior for everyone without toolchains having to specify non-default options. For example, I'm wondering whether it makes sense to limit the multi-use case to FP ops only, but not integers ops?

thetheodor commented:
> Hi @thetheodor, Yep, limiting the multi-use case to FP/vector ops would still work.

Hello, just to clarify, do you mean the following? 
- FP: allow multi-use
- Vector Int: allow multi-use
- Scalar Int: disallow multi-use

thetheodor commented:
Based on feedback I made the following changes:
- Removed the proposed flag
- Limited the multi-use case to FP and Vector arithmetic
- Adjusted  most test cases in `hoist-binop.ll`:
   - I removed the second uses in the `add` cases to allow the transformation.
   - Converted most `mul` cases to vector arithmetic, the reason is two-fold: 1) to test that vector arithmetic with multiple cases is allowed and 2) if I simply remove the second use `call void @use` then a different transformation is triggered (`hoistMulAddAssociation`)
   - Dropped a few single-use cases that trigger `hoistMulAddAssociation` instead of `hoistBOAssociation`
   
  
  @nikic @rj-jesus could you help with reviewing this version?

rj-jesus commented on llvm/test/Transforms/LICM/hoist-binop.ll:
```diff
@@ -432,35 +425,10 @@ entry:
 
-;
-; Hoist MUL and drop NSW even if both ops have it.
-define void @mul_no_nsw_2(i64 %c1, i64 %c2) {
```
Was removing this test intentional?

thetheodor replied:
The reason I removed it is that it no longer tests `hoistBOAssociation`, but it triggers `hoistMulAddAssociation`:

```
define void @mul_no_nsw_2(i64 %c1, i64 %c2) {
entry:
  %factor.op.mul = mul i64 %c1, %c2
  br label %loop

loop:                                             ; preds = %loop, %entry
  %index = phi i64 [ 0, %entry ], [ %step.add.reass, %loop ]
  %step.add.reass = mul i64 %index, %factor.op.mul
  br label %loop
}
```

(the vector version behaves similarly)

rj-jesus replied:
What if the type's changed to `<2 x i64>`?

thetheodor replied:
But `mul_one_use`:

scalar;
```
define void @mul_one_use(i64 %c1, i64 %c2) {
entry:
  %factor.op.mul = mul i64 %c1, %c2
  br label %loop

loop:                                             ; preds = %loop, %entry
  %index = phi i64 [ 0, %entry ], [ %step.add.reass, %loop ]
  %step.add.reass = mul i64 %index, %factor.op.mul
  br label %loop
}
```
vector:
```
define void @mul_one_use(<2 x i64> %c1, <2 x i64> %c2) {
entry:
  %factor.op.mul = mul <2 x i64> %c1, %c2
  br label %loop

loop:                                             ; preds = %loop, %entry
  %index = phi <2 x i64> [ zeroinitializer, %entry ], [ %step.add.reass, %loop ]
  %step.add.reass = mul <2 x i64> %index, %factor.op.mul
  br label %loop
}
```

rj-jesus reviewed (commented):
Thanks for the changes. Could you please regenerate `llvm/test/Transforms/LICM/hoist-binop.ll` with `--reset-variable-names`? I think that should simplify some of the variable names like `CVEC1_INIT` in `@mul_nuw`.

Also, it looks like `llvm/test/Transforms/LICM/sink-foldable.ll` and a couple of PowerPC tests might need updating.

rj-jesus commented on llvm/test/Transforms/LICM/hoist-binop.ll:
```diff
@@ -409,59 +405,56 @@ entry:
-; CHECK-LABEL: @add_both_nsw_second_nuw(
+;
+; Hoist MUL and drop NSW even if both ops have it.
+define void @mul_no_nsw_2(<2 x i64> %c1, <2 x i64> %c2) {
```
Could you move this below `@add_both_nsw_second_nuw` (where I believe it used to be before) to simplify the diff?

rj-jesus reviewed (approved):
LGTM, cheers (please allow a couple of days before merging in case anyone else has other comments).

nikic reviewed (approved):
LGTM

thetheodor commented:
> LLVM Buildbot has detected a new failure on builder `openmp-s390x-linux` running on `systemz-1` while building `llvm` at step 6 "test-openmp".
> 
> Full details are available at: https://lab.llvm.org/buildbot/#/builders/88/builds/15045
> 
> Here is the relevant piece of the build log for the reference

This must be unrelated to my change. I tried to reproduce it locally and couldn't.
